File: packages/contracts-python/src/aegis_contracts/blast_radius.py
"""Containment blast-radius preview contracts (pre-approval decision support).

A deterministic projection of the collateral of a Class 2/3 containment action, computed by
graph traversal over the run's current operator-visible graph snapshot — no LLM. Before an
operator (or a BASTION-proposal approver) commits, they see the severed/degraded edges, the
directly impacted neighbour assets, and the downstream service chain the action would touch.
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, model_validator

from aegis_contracts.entities import ActionClass
from aegis_contracts.errors import ContractErrorCode, ContractValidationError
from aegis_contracts.primitives import AssetId, RunId
from aegis_contracts.proposals import ScenarioCommandTemplateV1
from aegis_contracts.versioning import (
    BLAST_RADIUS_PREVIEW_SCHEMA_VERSION,
    assert_supported_schema_version,
)

__all__ = [
    "BlastRadiusDownstreamV1",
    "BlastRadiusImpactKindV1",
    "BlastRadiusImpactV1",
    "BlastRadiusPreviewV1",
]


class BlastRadiusImpactKindV1(StrEnum):
    """How a containment action touches a neighbouring edge/asset."""

    SEVERED = "severed"
    DEGRADED = "degraded"
    OUTAGE = "outage"


class BlastRadiusImpactV1(BaseModel):
    """A single 1-hop neighbour affected by the action, via one graph edge."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    asset_id: AssetId = Field(alias="assetId")
    label: str
    criticality: float = Field(ge=0.0, le=1.0)
    status: str
    relationship_type: str = Field(alias="relationshipType")
    impact_kind: BlastRadiusImpactKindV1 = Field(alias="impactKind")
    edge_id: str = Field(alias="edgeId")


class BlastRadiusDownstreamV1(BaseModel):
    """A service reached transitively through the dependency chain from the target."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    asset_id: AssetId = Field(alias="assetId")
    label: str
    criticality: float = Field(ge=0.0, le=1.0)
    hops: int = Field(ge=1)


class BlastRadiusPreviewV1(BaseModel):
    """Projected collateral of a containment (command, target) over the current graph."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    run_id: RunId = Field(alias="runId")
    command: ScenarioCommandTemplateV1
    target_asset_id: AssetId = Field(alias="targetAssetId")
    action_class: ActionClass = Field(alias="actionClass")
    severed_edge_count: int = Field(alias="severedEdgeCount", ge=0)
    degraded_edge_count: int = Field(alias="degradedEdgeCount", ge=0)
    impacted_assets: list[BlastRadiusImpactV1] = Field(
        alias="impactedAssets", default_factory=list
    )
    downstream_assets: list[BlastRadiusDownstreamV1] = Field(
        alias="downstreamAssets", default_factory=list
    )
    warnings: list[str] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_schema_version(self) -> BlastRadiusPreviewV1:
        assert_supported_schema_version("blast_radius_preview", self.schema_version)
        if self.schema_version != BLAST_RADIUS_PREVIEW_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported blast radius preview schema: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self

File: packages/contracts-ts/src/blast-radius.ts
import { z } from 'zod';

import { assetIdSchema, runIdSchema } from './primitives';
import { scenarioCommandTemplateSchema } from './proposals';
import { BLAST_RADIUS_PREVIEW_SCHEMA_VERSION } from './versioning';

const schemaVersionCheck = (expected: number) =>
  z
    .number()
    .int()
    .min(1)
    .superRefine((value, ctx) => {
      if (value !== expected) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: `Unsupported schema version: ${String(value)}`,
        });
      }
    });

const actionClassSchema = z.enum(['class_0', 'class_1', 'class_2', 'class_3']);

export const blastRadiusImpactKindSchema = z.enum(['severed', 'degraded', 'outage']);

export type BlastRadiusImpactKindV1 = z.infer<typeof blastRadiusImpactKindSchema>;

// A single 1-hop neighbour affected by a containment action, via one graph edge.
export const blastRadiusImpactSchema = z
  .object({
    assetId: assetIdSchema,
    label: z.string(),
    criticality: z.number().min(0).max(1),
    status: z.string(),
    relationshipType: z.string(),
    impactKind: blastRadiusImpactKindSchema,
    edgeId: z.string(),
  })
  .strict();

export type BlastRadiusImpactV1 = z.infer<typeof blastRadiusImpactSchema>;

// A service reached transitively through the dependency chain from the target.
export const blastRadiusDownstreamSchema = z
  .object({
    assetId: assetIdSchema,
    label: z.string(),
    criticality: z.number().min(0).max(1),
    hops: z.number().int().min(1),
  })
  .strict();

export type BlastRadiusDownstreamV1 = z.infer<typeof blastRadiusDownstreamSchema>;

// Projected collateral of a containment (command, target) over the current graph.
export const blastRadiusPreviewSchema = z
  .object({
    schemaVersion: schemaVersionCheck(BLAST_RADIUS_PREVIEW_SCHEMA_VERSION),
    runId: runIdSchema,
    command: scenarioCommandTemplateSchema,
    targetAssetId: assetIdSchema,
    actionClass: actionClassSchema,
    severedEdgeCount: z.number().int().min(0),
    degradedEdgeCount: z.number().int().min(0),
    impactedAssets: z.array(blastRadiusImpactSchema).default([]),
    downstreamAssets: z.array(blastRadiusDownstreamSchema).default([]),
    warnings: z.array(z.string()).default([]),
  })
  .strict();

export type BlastRadiusPreviewV1 = z.infer<typeof blastRadiusPreviewSchema>;
